### ISSUE TITLE
FEATURE REQUEST Android TV

### ISSUE THREAD

**Issue by popeye2468** [jarnedemeulemeester/findroid#23]
Hi please could you make this client for Android TV I love how fast this client is, it's unbeatable 

So again Please for Android TV

**Comment by Taihouu** [comment]
This would be extremely appreciated since it seems to support far more codecs and the current Android TV application is a bit of a mess, especially on devices like the Shield where a lot of content is transcoded when it shouldn't be.

**Comment by Gottarocket** [comment]
I know its for android tv but android tv can allow apk to be installed which works on mobile this maynt be the case here but most do but it would be helpful if you do build it for  webos too as there is no option to sideload thek on our end

**Comment by krutburken** [comment]
Yes please :)

**Comment by krutburken** [comment]
@jarnedemeulemeester Maybe I'm just being a bit dumb here, but how do I switch to the android-tv layout?. Saw it was included in 0.3.0 and wanted to try it out.
I sideloaded the apk to my device, but from what I can tell it looks the same as before

**Comment by jarnedemeulemeester** [comment]
> @jarnedemeulemeester Maybe I'm just being a bit dumb here, but how do I switch to the android-tv layout?. Saw it was included in 0.3.0 and wanted to try it out. I sideloaded the apk to my device, but from what I can tell it looks the same as before

When I sideload the apk to my emulator and open it it shows the tv layout. Maybe try force closing the app. 
Also note that the tv layout is not really usable yet, but you can try it if you want.

**Comment by krutburken** [comment]
@jarnedemeulemeester I've tried it on both a Shield TV and Google TV without success. Maybe I'll try it in a emulator to see if it behaves differently

**Comment by mind-overflow** [comment]
hey! just wanted to chime in, is there any update about the tv version? :)

**Comment by jarnedemeulemeester** [comment]
I have started to fix the TV version.

I have also bought an Android TV which will arrive soon. 
That should also push me to work on the TV version :smile:

**Comment by mind-overflow** [comment]
hello, just commenting to see if there's any news about the android tv version!

**Comment by fmchris10** [comment]
I actually have no end of issues with Jellyfin's official apps: Findroid offers smooth play back. Whilst it's currently usable, I do think there are areas that would naturally improve it for Android TV users.

**Comment by jarnedemeulemeester** [comment]
Let me give a quick update:
Right now the code of the TV version is interwoven in the main (phone) version. Which is good for reusing a lot of things but not for code maintainability and readability.
The obvious solution is to split the whole app into modules. Also called "app modularization". 
The Phone and TV modules can then use other modules they need. Like a core module and data module etc.
This will also allow for a Wear OS version in the future which will only use a music module (so no video) etc.
The degree of modularization (layer vs feature vs layer & feature) is not finalized yet.

TLDR I will have to completely refactor the app to be able to continue working on the TV version without making everything a complete mess. And that's what I'll be focusing on right now.

Sorry everyone for the long wait.

**Comment by Rydberg95** [comment]
Doing great with the app, and I prefer it over the official one. 

I've been experimenting a bit with different players to get direct play (with HW decoding) to work on Android TV (Chromecast latest gen). The file I've tried with is a 20GB-ish HEVC 4k one, and I've come to some conclusions. Not sure if you have any use for them or if it's obvious things for you, but it here goes:

- Playback does not work with default (exo?) player.
- Playback is choppy but works with mpv. The video and sound becomes increasingly out of sync during the chopping.
- Playback works well when direct playing the file in MX player, with their custom codecs imported (easily found when searching) and "HW+" enabled.

So, this makes me believe that the actual hardware in the device is perfectly capable of playing, but its "just" a matter of decoding. 

Here i really don't know what I'm talking about, but: maybe the same codecs can implemented in droidfin, of that optional "import" of codecs can take place (just as in MX player)?

Keep up the good work and thanks the app!

**Comment by jarnedemeulemeester** [comment]
@Rydberg95 Thank you for the extra information. 

I don't think Findroid can use MX player codecs and I am already focussing on mpv right now. 
Speaking of mpv see below a new APK with an updated version of libmpv.

Also here are my plans for Android TV:
I will remove all Android TV code from Findroid for the next release. But don't worry this will be added again later with a completely new design in a separate module. My goal is to add it again this year.

To compensate for the lack of Android TV in the next release, I will provide this APK which still has the Android TV code and updated mpv with hardware AV1 decoding (if the device supports it).

[findroid-armeabi-v7a-release.apk](https://drive.google.com/file/d/1fpseQesqHaFL4TKD41SfOVSHhNImY2wh/view?usp=sharing)

Also try setting Hardware decoding to `mediacodec`. This is the only setting which allows my Sony TV to play videos.

**Comment by Rydberg95** [comment]
@jarnedemeulemeester 
I suspected it wasn't as easy as just implementing codecs...

Awesome with the new release and AV1 decoding! I downloaded it and tried with my file, but unfortunately with the same results. Great that you're working on it though!

**Comment by duomiantii** [comment]
@jarnedemeulemeester
My TV is Android 5.0, will it be supported in future versions of Android TV? My  TV can't install this app now.

**Comment by jarnedemeulemeester** [comment]
> My TV is Android 5.0, will it be supported in future versions of Android TV? My  TV can't install this app now.

The current minimum Android version is 8.1 and this will not go any lower. I am looking at possibly upping it to Android 9 this year. And I plan to keep the minimum on that version for quite some time.

So the TV version will likely release with a minimum requirement of Android 9.

**Comment by panvar** [comment]
Do you have any news on when you will release findroid for Android TV?

**Comment by Raik0** [comment]
Any update to android tv version? No rush, just thought I'd ask.

**Comment by nurailroaddev** [comment]
If you need testers, I have shield, shield pro, ccwgtv, mii2.

**Comment by jothi-prasath** [comment]
+1

**Comment by Nassr199** [comment]
I would like to know about the latest update to the Android TV version

**Comment by drizzt09** [comment]
Interested in this for sure. Only thing holding me back from officially moving away from Emby is a good firetv remote capable interface. Thanks

**Comment by jimmutter** [comment]
I can also volunteer for Shield testing.  I have both the normal and pro of the latest iteration.

**Comment by IGS-GIT** [comment]
I can volunteer for GoogleTV/BRAVIA if the need arises. I know Sony likes to make things difficult sometimes so maybe there's something I can help with.

**Comment by shbsky** [comment]
Any update to android tv version? No rush, just thought I'd ask.  +1

**Comment by EnduringGuerila** [comment]
I installed the v0.14.2-libre-armeabi-v7a.apk
on our Sony Bravia VH1. 
Playback seems to work great, fixed some issue we had with the official app. Navigating the interface is just terrible with the remote. I saw on here that there is a TV version that was already merged into main but hasn't been released yet? What's the status on that?

**Comment by tm9k1** [comment]
the code is in the tv folder but somebody needs to build it and provide an APK for us to test.

**Comment by GrabbenD** [comment]
@tm9k1 Compile it with Android Studio

APKs redistributed by random users might not be safe

**Comment by tm9k1** [comment]
I know, @GrabbenD , thanks for pointing it out for everyone though.

Since the nice team has enabled CI for findroid, I was able to grab a debug APK from https://github.com/jarnedemeulemeester/findroid/actions . I won't list down the steps because again, it's NOT a release, but a build for devs (and impatient people like myself) to see the direction in which the development is going. I am very happy to report the developments are on point, and I'm very eagerly looking forward to the TV build's release!

**Comment by danwilldev** [comment]
I have also just grabbed the apk from the CICD job. To view content on a Chromecast with Google TV. Enable experimental MPV. And set the options to mediacodec-copy, GPU and opensles. 
I'll eagerly await development on this. But it's cool to see the groundwork already laid out.

**Comment by drizzt09** [comment]
> I have also just grabbed the apk from the CICD job. To view content on a Chromecast with Google TV. Enable experimental MPV. And set the options to mediacodec-copy, GPU and opensles. 
> I'll eagerly await development on this. But it's cool to see the groundwork already laid out.

How do I grab the latest and build it in studio myself. Would like to play with the TV variant.

**Comment by drizzt09** [comment]
> > APKs redistributed by random users might not be safe
> 
> I know, @GrabbenD , thanks for pointing it out for everyone though.
> 
> Since the nice team has enabled CI for findroid, I was able to grab a debug APK from https://github.com/jarnedemeulemeester/findroid/actions . I won't list down the steps because again, it's NOT a release, but a build for devs (and impatient people like myself) to see the direction in which the development is going. I am very happy to report the developments are on point, and I'm very eagerly looking forward to the TV build's release!

@tm9k1 can you DM me steps please. Would like to try it out.

**Comment by anvishinc** [comment]
@jarnedemeulemeester not sure if this is a known bug or something intentional, but out of curiosity and frustration with the stock jellyfin Android tv client I tried installing the findroid android tv app from ci build of the latest release 0.15.1 and it wouldn't go beyond the server url entry screen, the connect button seemed to be disabled

**Comment by jarnedemeulemeester** [comment]
> @jarnedemeulemeester not sure if this is a known bug or something intentional, but out of curiosity and frustration with the stock jellyfin Android tv client I tried installing the findroid android tv app from ci build of the latest release 0.15.1 and it wouldn't go beyond the server url entry screen, the connect button seemed to be disabled

You are the second person I know who has this issue. The problem is not that the Connect button is disabled (because it's not), it's that for some reason you cannot move the selection from the input field to the button. 
Could you tell me what device you are running the tv build on?

**Comment by anvishinc** [comment]
> > @jarnedemeulemeester not sure if this is a known bug or something intentional, but out of curiosity and frustration with the stock jellyfin Android tv client I tried installing the findroid android tv app from ci build of the latest release 0.15.1 and it wouldn't go beyond the server url entry screen, the connect button seemed to be disabled
> 
> You are the second person I know who has this issue. The problem is not that the Connect button is disabled (because it's not), it's that for some reason you cannot move the selection from the input field to the button. 
> Could you tell me what device you are running the tv build on?

@jarnedemeulemeester I think you are right, I did try a few minutes back to use a keyboard but still couldn't move forward, so what you say definitely is the reason.

As for the device I am using, it is a Realme TV Stick 4k and it is running Android 11. And I used the armeabi_v7a apk from the git actions for 0.15.1 release.

Let me know if you need any more details or if I can help troubleshoot this issue in any way

**Comment by p0358** [comment]
So I used the sideloaded APK from Actions on a TV, and here's a few bugs/UI issues I noticed:
- on the login screen, pressing the submitting arrow on the on-screen keyboard does nothing (while it should move to the next element, so from server input to button; then from login to password, from password to button etc.); worth noting that I could dismiss the keyboard with Back and then navigate to the button with Down
- on the main screen, going up to tabs always highlights the first tab, instead of the currently active tab
- navigating around playback controls feels pretty finicky and non-intuitive, it should be like on main app that by default pressing OK is play/pause, can scrub with Left/Right, and then go Up for the extra buttons
- selecting different audio/subtitles track should dismiss the dialog instead of having to explicitly press Back after ticking the option of interest most likely

But these are minor things!!! Other than that, I immediately ticked on mpv player in the settings and I can easily play all the media files that the official app would commit sudoku upon. This is brilliant, imo it could already be very usable if just a couple of these issues are fixed as some kind of beta, it'd be already a superior experience in many ways. I might actually be putting up with these bugs and using this instead of the official app instead... (at least when the official one fails to actually play files (speaking of which it's beyond me why wouldn't they just use some underlying player that works))

**Comment by damomato** [comment]
I appreciate your work on this and am very excited for the tv version! I might finally move on from Kodi

**Comment by badaz** [comment]
Looking forward to the TV version as well
